substrate/client/network/src/ipfs_block_provider.rs

```diff
@@ -0,0 +1,305 @@
+		this.fetched_to = this.fetched_to.max(pruned_to); // Don't try to fetch pruned blocks!
+		while let (only_block, Some(block)) = (this.blocks.len() == 1, this.blocks.front_mut()) {
+			if block.number >= pruned_to {
+				break // Not pruned
+			}
+
+			// Discard any transaction hashes that we didn't even add yet
+			if let (true, Some(added_to)) = (only_block, this.added_to) {
+				block.transaction_hashes.truncate(added_to);
+				this.added_to = None;
+			}
+
+			while let Some(hash) = block.transaction_hashes.pop() {
+				match this.extra_refs.entry(hash) {
+					Entry::Occupied(mut entry) => match entry.get().checked_sub(1) {
+						Some(extra_refs) => {
+							entry.insert(extra_refs);
+						},
+						None => {
+							entry.remove();
+							// TODO: don't assume hash is BLAKE2b, use `Multihash` instead.
+							return Poll::Ready(Some(Change::Removed(to_multihash::<BlakeTwo256>(
+								&hash,
+							))))
+						},
+					},
+					// This should not be possible!
+					Entry::Vacant(_) => warn!("Pruned transaction hash {hash} not found"),
+				}
+			}
+
+			this.blocks.pop_front();
+		}
+
+		// Handle finalized blocks
+		loop {
+			// Fetch finalized blocks
+			while this.added_to.is_none() && (this.fetched_to < this.finalized_to) {
+				let hashes = this.client.block_hash(this.fetched_to).and_then(|hash| {
+					let hash = hash.ok_or_else(|| {
+						sp_blockchain::Error::UnknownBlock(format!(
+							"Hash of block {} not found",
+							this.fetched_to,
+						))
+					})?;
+					this.client.block_indexed_hashes(hash)
+				});
+				match hashes {
+					Ok(Some(hashes)) if !hashes.is_empty() => {
+						this.blocks.push_back(IndexedBlock {
+							number: this.fetched_to,
+							transaction_hashes: hashes,
+						});
+						this.added_to = Some(0);
+					},
+					Ok(_) => (),
+					Err(err) => debug!("Error fetching block {}: {err}", this.fetched_to),
+				}
+				this.fetched_to += 1u32.into();
+			}
+
+			// Add from last fetched block
+			while let Some(added_to) = &mut this.added_to {
+				let block = this.blocks.back().expect(
+					"added_to only set to Some after pushing a block, \
+					set to None before popping last block",
+				);
+				let hash = block.transaction_hashes[*added_to];
+				*added_to += 1;
+				if *added_to == block.transaction_hashes.len() {
+					this.added_to = None;
+				}
+
+				match this.extra_refs.entry(hash) {
+					Entry::Occupied(mut entry) => *entry.get_mut() += 1,
+					Entry::Vacant(entry) => {
+						entry.insert(0);
+						// TODO: don't assume hash is BLAKE2b, use `Multihash` instead.
+						return Poll::Ready(Some(Change::Added(to_multihash::<BlakeTwo256>(&hash))))
+					},
+				}
+			}
+
+			// Fully handled last fetched block. Loop if there are more blocks to fetch, otherwise
+			// nothing to do.
+			debug_assert!(this.fetched_to <= this.finalized_to);
+			if this.fetched_to == this.finalized_to {
+				return Poll::Pending
```

Comment:
Since we are not saving the waker, we rely entirely on `this.finality_notifications.poll_next_unpin(cx)` to wake us up, is that intended?